File: Typescript/src/01.AlgorithmicToolbox/Week1ProgrammingChallenges/MaxPairwiseProduct.test.ts
import { getMaxPair, maxPairwiseProduct } from "./MaxPairwiseProduct"

describe("max pairwise product", () => {
  const numbers = [21468, 16859, 82178, 70496, 82939, 44491]

  it("get max pair returns largest two numbers", () => {
    const maxPair = getMaxPair(numbers)
    expect(maxPair).toContain(82178)
    expect(maxPair).toContain(82939)
  })
  it("max pairwise product returns correct product", () => {
    expect(maxPairwiseProduct(numbers)).toBe(6815761142)
  })
})

File: Elixir/lib/01.AlgorithmicToolbox/Week1ProgrammingChallenges/max_pairwise_product_test.exs
defmodule MaxPairwiseProductTest do
  use ExUnit.Case

  import DataStructuresAndAlgorithms.AlgorithmicToolBox.Week1ProgrammingChallenges.MaxPairwiseProduct

  doctest DataStructuresAndAlgorithms.AlgorithmicToolBox.Week1ProgrammingChallenges.MaxPairwiseProduct

  test "returns correct max pairwise product" do
    assert max_pairwise_product([21468, 16859, 82178, 70496, 82939, 44491]) ==
             {:ok, 6_815_761_142}
  end

  test "returns error on invalid input" do
    assert max_pairwise_product([-100, 2, 3]) == {:error, :invalid_input}
  end
end
